formatters/critical-request-chains.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const url = require('url');
+const path = require('path');
+const fs = require('fs');
+const Formatter = require('./formatter');
+const html = fs.readFileSync(path.join(__dirname, 'partials/critical-request-chains.html'), 'utf8');
+
+class CriticalRequestChains extends Formatter {
+
+  /**
+   * gets the formatter for the CLI Printer and the HTML report.
+   */
+  static getFormatter(type) {
+    switch (type) {
+      case 'pretty':
+        return function(info) {
+          const longestChain = CriticalRequestChains._getLongestChainLength(info);
+          const longestDuration = CriticalRequestChains._getLongestChainDuration(info);
+          const urlTree = CriticalRequestChains._createURLTreeOutput(info);
+
+          const output = `    - Longest request chain (shorter is better): ${longestChain}\n` +
+          `    - Longest chain duration (shorter is better): ${longestDuration.toFixed(2)}ms\n` +
+          '    - Initial navigation\n' +
+              '      ' + urlTree.replace(/\n/g, '\n      ') + '\n';
+          return output;
+        };
+
+      case 'html':
+        // Returns a handlebars string to be used by the Report.
+        return html;
+
+      default:
+        throw new Error('Unknown formatter type');
+    }
+  }
+
+  static _traverse(tree, cb) {
+    function walk(node, depth, startTime) {
+      const children = Object.keys(node);
+      if (children.length === 0) {
+        return;
+      }
+
+      children.forEach(id => {
+        const child = node[id];
+        if (!startTime) {
+          startTime = child.request.startTime;
+        }
+
+        // Call the callback with the info for this child.
+        cb({
+          depth,
+          id,
+          node: child,
```

Comment:
if we just call it node here and rename the walk parent param to not be node this gets a lot nicer